### PR TITLE
[FIX] base: allow merged contacts to retain product link access in emails

### PR DESCRIPTION
**Pull request by walidsahli**
Problem:
While merging two partners in the contact app using the _update_foreign_keys method, the foreign keys are merged using SQL instead of ORM. The issue is that the partner_share field is computed based on the user_ids field and is stored. If we update using SQL, it will not trigger the _compute_partner_share method, which will set the latest value.

Other problems might arise for fields that are computed based on foreign keys updated using SQL during a merge.

Steps to reproduce:

In a new database, create a basic contact (individual, with name and email).
Merge this contact with another contact who has document access (e.g., Marc Demo).
Go to any product page.
Add a log note and tag the new contact.
The contact will receive an email without the "View product" button.

opw-3864317

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by nle-odoo — COMMENTED** [review]
I've not test the given use case, but in my test the user_ids of the partner all ends up on the target partner, it seems to be done two lines above this PR fix:

https://github.com/odoo/odoo/blob/1d2705ebe641ba91a8487fde8e04994812e99f5e/odoo/addons/base/wizard/base_partner_merge.py#L317

in a debugger:
```py
# before this line
--> 318         self._update_foreign_keys(src_partners, dst_partner)
> dst_partner.user_ids
res.users()
# after this line
--> 319         self._update_reference_fields(src_partners, dst_partner)
res.users(12,)
```
Possibly your change fix the issue, but I don't understand how since we are doing in ORM what should have already be done in SQL. If the issue is resolved with your fix, we should check why this doesn't work by itself and if there is some particularity in the generic code (in method `_update_foreign_keys`) that should be fixed.

**Comment by walidsahli** [comment]
> I've not test the given use case, but in my test the user_ids of the partner all ends up on the target partner, it seems to be done two lines above this PR fix:
> 
> https://github.com/odoo/odoo/blob/1d2705ebe641ba91a8487fde8e04994812e99f5e/odoo/addons/base/wizard/base_partner_merge.py#L317
> 
> in a debugger:
> 
> ```python
> # before this line
> --> 318         self._update_foreign_keys(src_partners, dst_partner)
> > dst_partner.user_ids
> res.users()
> # after this line
> --> 319         self._update_reference_fields(src_partners, dst_partner)
> res.users(12,)
> ```
> 
> Possibly your change fix the issue, but I don't understand how since we are doing in ORM what should have already be done in SQL. If the issue is resolved with your fix, we should check why this doesn't work by itself and if there is some particularity in the generic code (in method `_update_foreign_keys`) that should be fixed.

Thank you. It makes perfect sense, and after further investigation, it turns out that there was another problem, but the solution remains the same. I updated the PR description with more information about it.

**Review by nle-odoo — APPROVED** [review]
current commit version makes sense to me. if we update the user_ids in SQL, we needs to update the partner_share that relies on `@api.depends('user_ids.share', 'user_ids.active')`

note: as said in runbot, you should use eg. `dst_partner._recompute_recordset(['partner_share'])` instead of deprecated `recompute`

File: odoo/addons/base/wizard/base_partner_merge.py [hunk 1]
```diff
@@ -318,6 +318,9 @@ def _merge(self, partner_ids, dst_partner=None, extra_checks=True):
         self._update_reference_fields(src_partners, dst_partner)
         self._update_values(src_partners, dst_partner)
 
+        self.env.add_to_compute(dst_partner._fields['partner_share'], dst_partner)
+        dst_partner._recompute_recordset(['partner_share'])
```
Comment (ryv-odoo) [inline odoo/addons/base/wizard/base_partner_merge.py:322]: This line looks useless to me. Every record tags to recompute will be recompute at the end of request. No need to call it yourself.

File: odoo/addons/base/wizard/base_partner_merge.py [hunk 2]
```diff
@@ -318,6 +318,9 @@ def _merge(self, partner_ids, dst_partner=None, extra_checks=True):
         self._update_reference_fields(src_partners, dst_partner)
         self._update_values(src_partners, dst_partner)
 
+        self.env.add_to_compute(dst_partner._fields['partner_share'], dst_partner)
```
Comment (ryv-odoo) [inline odoo/addons/base/wizard/base_partner_merge.py:321]: Not against this specific change but normally we have a specific method to find what we need to recompute depending of the change done on the recordset: `modified([<field_change>])` (check write by example).

**Review by ryv-odoo — COMMENTED** [review]
Hello @walidsahli ,

Nice catch.

Note that the code of this wizard seems very old. And I am pretty sure that others inconsistencies can be found due to the SQL request done and because the ORM is not aware of the change done in DB.

**Comment by ryv-odoo** [comment]
@robodoo r+
LGTM.

**Pull request closed without merging**